reactor-core/src/main/java/reactor/core/publisher/ContextTrackingFunctionWrapper.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2019-Present Pivotal Software Inc, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.core.publisher;
+
+import java.util.function.Function;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import reactor.core.CorePublisher;
+import reactor.core.CoreSubscriber;
+import reactor.core.publisher.FluxContextStart.ContextStartSubscriber;
+import reactor.util.context.Context;
+
+/**
+ * This {@link Function} wrapper is used by reactor-tools to implement the context loss detection.
+ *
+ */
+@SuppressWarnings("unused")
+class ContextTrackingFunctionWrapper<T, V> implements Function<CorePublisher<T>, CorePublisher<V>> {
+
+	static final String CONTEXT_MARKER_PREFIX = "reactor.core.context.marker.";
+
+	final Function<? super Publisher<T>, ? extends Publisher<V>> transformer;
+
+	ContextTrackingFunctionWrapper(Function<? super Publisher<T>, ? extends Publisher<V>> transformer) {
+		this.transformer = transformer;
+	}
+
+	@Override
+	public CorePublisher<V> apply(CorePublisher<T> self) {
+		String key = CONTEXT_MARKER_PREFIX + System.identityHashCode(self);
+
+		Publisher<V> newSource = Operators.<T, T>liftPublisher((p, actual) -> {
+			Context ctx = actual.currentContext();
+
+			if (!ctx.hasKey(key)) {
+				throw new IllegalStateException("Context loss after applying " + transformer);
+			}
+
+			Context newContext = ctx.delete(key);
+			return new ContextStartSubscriber<>(actual, newContext);
+		}).andThen(transformer).apply(self);
+
+		return new CorePublisher<V>() {
```

Comment:
for `newSource`, type is maintained by `liftPublisher`. But here you are returning a `Publisher` where there was potentially a `Flux` or `Mono`. Isn't that problematic?

Reply:
in `transform`, there is `from` on the function's returned value anyways, so that it looks like this:
```java
Mono<V> transform(fn) {
    fn = new Wrapper(fn);
    return onAssembly(from(fn.apply(this)));
}
```

Reply:
ah, got it. please add a comment for our future selves 😛